### PR TITLE
Exposed rng generation device and dtype for augmentations.

### PR DESCRIPTION
**Pull request by shijianjian**
### Description

Inspired by #762.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: test/augmentation/test_random_generator.py [hunk 1]
```diff
@@ -97,38 +98,39 @@ def test_valid_param_combinations(
...
-        pytest.param(None, None, torch.tensor(0.), None, marks=pytest.mark.xfail),
-        pytest.param(None, None, torch.tensor([0.]), None, marks=pytest.mark.xfail),
-        pytest.param(None, None, torch.tensor([0., 1., 2.]), None, marks=pytest.mark.xfail),
-
-        pytest.param(None, None, None, torch.tensor([-1., 0.]), marks=pytest.mark.xfail),
-        pytest.param(None, None, None, torch.tensor([0, 1.]), marks=pytest.mark.xfail),
-        pytest.param(None, None, None, torch.tensor(0.), marks=pytest.mark.xfail),
-        pytest.param(None, None, None, torch.tensor([0.]), marks=pytest.mark.xfail),
-        pytest.param(None, None, None, torch.tensor([0., 1., 2.]), marks=pytest.mark.xfail),
+        pytest.param(torch.tensor([-1., 2.]), None, None, None),
```
Comment (edgarriba) [inline test/augmentation/test_random_generator.py:101]: not quite sure now the point of having `pytest.param`

**Review by edgarriba — APPROVED** [review]
LGTM. Only the `pytest.param` thing. Besides, have you tested on cuda?

**Comment by shijianjian** [comment]
@edgarriba In cuda/ cpu/tpu, ```Uniform().rsample``` yields different results under the same seed.

I have filed an issue https://github.com/pytorch/pytorch/issues/47347.

Probably, we may only ensure the deterministic with CPU only and skip the value check with GPU/TPUs.

**Comment by shijianjian** [comment]
@edgarriba In cuda/ cpu/tpu, ```Uniform().rsample``` yields different results under the same seed.

I have filed an issue https://github.com/pytorch/pytorch/issues/47347.

Probably, for now, we may only ensure the deterministic with CPU only and skip the value check with GPU/TPUs.

**Comment by shijianjian** [comment]
We have to make a choice there:
1. To put all random param calculations into CPU, to push the tensor from devices to devices.
2. Only to ensure the CPU deterministically, adding skip to GPU/TPU. For GPU/TPU, we only to check if they can run with valid parameters.

Personally, I would say the second option.

**Comment by edgarriba** [comment]
@shijianjian not sure if I agree with approach 2, which means that running the augmentations in different device might lead to different solutions. /cc @pmeier @ducha-aiki @FGeri What do you think here ?

**Comment by FGeri** [comment]
> We have to make a choice there:
> 
> 1. To put all random param calculations into CPU, to push the tensor from devices to devices.
> 2. Only to ensure the CPU deterministically, adding skip to GPU/TPU. For GPU/TPU, we only to check if they can run with valid parameters.
> 
> Personally, I would say the second option.

@edgarriba @shijianjian How about monkeypatching the random sampling functions to generate on the cpu for the tests? Then no performance drop during runtime but still can ensure correct operation on every device.

**Comment by edgarriba** [comment]
I think that providing different behavior based on your hardware configuration might lead to complex debugging and reducing the reproducibility ability, at least from user perspective. One use case is prototyping your code in CPU where you can a set of suitable parameters for your final training. In case that the GPU/TPU produce different results, the parameters search time will eventually grow.

**Comment by anguelos** [comment]
Since pytorch doesn't guaranty cross-device determinism of the seed, even CPU does not guaranty reproducibility. I suspect ARM vs Intel might differ and possibly MLK, OpenBLAS or other hardware accelerated libraries might behave differently even in CPU. That is why in my opinion you should look for consistency and consider that testcases based on torch.Distributions should be tested statistically and with margins of error.

**Comment by shijianjian** [comment]
I think @anguelos has proposed the right way to test the distribution samples (though hard to be implemented 😂).

I do not think the searching time will be increased since it is just using different random generators, while still the same distribution. Whilst testing on CPU, we normally only test the edge cases to determine the hyperparams, not the sampled params. It should be fine.

So, the problem is DO WE NEED TO REPRODUCE?

**Comment by anguelos** [comment]
Just one thing: In the case noise augmentations (eg: gaussian noise) you might need one sample from the distribution per pixel. That can be quite expensive.

**Comment by edgarriba** [comment]
@shijianjian what's the status/decision for this ?

**Comment by shijianjian** [comment]
I have added ```set_rng_device_and_dtype``` in augmentation base class. By default, all numerical tests will be running on CPU and the default dtype, while they can be updated if the user calls ```set_rng_device_and_dtype```.

File: kornia/augmentation/random_generator/random_generator.py [hunk 2]
```diff
@@ -263,52 +333,58 @@ def random_crop_generator(
             If tensor, it must be (B, 2).
         resize_to (tuple): Desired output size of the crop, like (h, w). If None, no resize will be performed.
         same_on_batch (bool): apply the same transformation across the batch. Default: False.
+        device (torch.device): the device on which the random numbers will be generated. Default: cpu.
+        dtype (torch.dtype): the data type of the generated random numbers. Default: float32.
 
     Returns:
         params Dict[str, torch.Tensor]: parameters to be passed for transformation.
+            - src (tensor): cropping bounding boxes with a shape of (B, 4, 2).
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/random_generator.py:341]: `torch.Tensor` ?

File: kornia/augmentation/utils/helpers.py [hunk 3]
```diff
@@ -226,51 +226,60 @@ def _adapted_uniform(
...
+    By default, sampling happens on the default device and dtype. If low/high is a tensor, sampling will happen
+    in the same device/dtype as low/high tensor.
     """
     device, dtype = _extract_device_dtype([
         low if isinstance(low, torch.Tensor) else None,
         high if isinstance(high, torch.Tensor) else None,
     ])
     if not isinstance(low, torch.Tensor):
-        low = torch.tensor(low, device=device, dtype=torch.float64)
+        low = torch.tensor(float(low), device=device, dtype=dtype)
```
Comment (edgarriba) [inline kornia/augmentation/utils/helpers.py:244]: I think in similar places like this one we could start using `torch.as_tensor`
https://pytorch.org/docs/stable/generated/torch.as_tensor.html?highlight=as_tensor#torch.as_tensor

we need to verify backward compatibility though

File: test/augmentation/test_augmentation.py [hunk 4]
```diff
@@ -448,6 +448,27 @@ def test_same_on_batch(self, device, dtype):
...
+              [0.8529, 0.9529, 1.0000]]],
+            [[[0.2660, 0.3660, 0.4660],
+              [0.7660, 0.6660, 0.5660],
+              [0.8660, 0.9660, 1.0000]],
+             [[0.2660, 0.3660, 0.4660],
+              [0.7660, 0.6660, 0.5660],
+              [0.8660, 0.9660, 1.0000]],
+             [[0.2660, 0.3660, 0.4660],
+              [0.7660, 0.6660, 0.5660],
+              [0.8660, 0.9660, 1.0000]]]], device=device, dtype=dtype)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:470]: `_get_expected_brightness` ?

File: test/augmentation/test_augmentation.py [hunk 5]
```diff
@@ -547,27 +538,31 @@ def test_random_contrast_list(self, device, dtype):
...
-                                   [0.4926, 0.4105, 0.3284],
-                                   [0.5747, 0.6568, 0.8210]],
-                                  [[0.0821, 0.1642, 0.2463],
-                                   [0.4926, 0.4105, 0.3284],
-                                   [0.5747, 0.6568, 0.8210]]]], device=device, dtype=dtype)
+        expected = self.expected_contrast(device, dtype)
 
         assert_allclose(f(input), expected, atol=1e-4, rtol=1e-5)
 
+    def expected_saturation(self, device, dtype):
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:545]: all this tests will change with https://github.com/kornia/kornia/pull/706, right ?

File: test/augmentation/test_random_generator.py [hunk 6]
```diff
@@ -97,38 +98,39 @@ def test_valid_param_combinations(
...
-        random_color_jitter_generator(
-            8,
-            brightness.to(device=device, dtype=dtype) if brightness is not None else None,
-            contrast.to(device=device, dtype=dtype) if contrast is not None else None,
-            saturation.to(device=device, dtype=dtype) if saturation is not None else None,
-            hue.to(device=device, dtype=dtype) if hue is not None else None
-        )
+        with pytest.raises(Exception):
+            random_color_jitter_generator(
+                8,
```
Comment (edgarriba) [inline test/augmentation/test_random_generator.py:128]: is this batch size ?

**Review by edgarriba — APPROVED** [review]
LGTM. Just quick questions. Did you test in a GPU ?

File: kornia/augmentation/augmentation3d.py [hunk 7]
```diff
@@ -555,14 +555,24 @@ class CenterCrop3D(AugmentationBase3D):
...
     Examples:
         >>> rng = torch.manual_seed(0)
-        >>> inputs = torch.randn(1, 1, 3, 3, 3)
+        >>> inputs = torch.randn(1, 1, 2, 4, 6)
+        >>> inputs
+        tensor([[[[[-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920],
+                   [-0.3160, -2.1152,  0.3223, -1.2633,  0.3500,  0.3081],
+                   [ 0.1198,  1.2377,  1.1168, -0.2473, -1.3527, -1.6959],
+                   [ 0.5667,  0.7935,  0.5988, -1.5551, -0.3414,  1.8530]],
+        <BLANKLINE>
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation3d.py:564]: For passing the doctest, the doctest checks the output string only, not the value. So it must be identical.

File: kornia/augmentation/utils/helpers.py (hunk 3 above)
Comment (shijianjian) [reply kornia/augmentation/utils/helpers.py:244]: Great. It should be okay for the earlier versions (it was there after 1.2.0).

File: test/augmentation/test_augmentation.py (hunk 5 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:545]: We need to update both sides. This PR will give more functionalities. #706 shall have more basic testings, like different device/dtype operations shall have different value outputs, etc.

File: test/augmentation/test_random_generator.py (hunk 6 above)
Comment (shijianjian) [reply test/augmentation/test_random_generator.py:128]: Yes
